lib/reek/examiner.rb

```diff
@@ -95,28 +97,24 @@ def smelly?
     # So we catch the exception here, let the user know something went wrong
     # and continue with the analysis.
     #
-    # @return an instance of Examiner
-    #
     # :reek:TooManyStatements: { max_statements: 6 }
     def run
-      @run ||= begin
-        syntax_tree = source.syntax_tree
-        return self unless syntax_tree
-        begin
-          examine syntax_tree
-        rescue StandardError => exception
-          $stderr.puts format(INCOMPREHENSIBLE_SOURCE_TEMPLATE, source.origin, exception.inspect)
-        else
-          smell_repository.report_on(collector)
-        end
-        self
+      syntax_tree = source.syntax_tree
+      return [] unless syntax_tree
+      begin
+        examine syntax_tree
+      rescue StandardError => exception
+        $stderr.puts format(INCOMPREHENSIBLE_SOURCE_TEMPLATE, source.origin, exception.inspect)
+        []
       end
     end
 
     def examine(syntax_tree)
+      result = []
       ContextBuilder.new(syntax_tree).context_tree.each do |element|
```

Comment:
How about writing this even conciser without the ugly initialization and explicit return below via:

```Ruby
def examine(syntax_tree)
  ContextBuilder.new(syntax_tree).context_tree.each_with_object([]) do |element, result|
  result += smell_repository.examine(element)
end
```

?